Add DetourAllocateRegionWithinJumpBounds API

msmania opened the pull request:
This patch exports the ability to allocate a region within the jump bounds from the given address because this ability is useful outside Detours.  The logic was already implemented as a part of the internal function
`detour_alloc_trampoline`.  The change is to extract it out as a public function `DetourAllocateRegionWithinJumpBounds`.

dtarditi commented:
Generally, we're trying to avoid expanding the functionality of Detours.  We want it to stay focused on the core problem that it solves.   This change seems reasonable, though, because it is simply opening an API as public.

I noticed when reviewing the code that ` DetourAllocateRegionWithinJumpBounds` tries to allocate a region that is a multiple of system page granularity (4 Kbytes).   This led to changes in the functions that it calls.   On Windows, the Virtual Memory subsystem won't allocate virtual memory regions at a granularity of less than 64 KBytes of memory anway.  Is there a reason why you are trying to allocate memory at a smaller granularity?

msmania commented:
I removed the second commit bringing 4K granularity part and rebased the branch.  Now the change is to export the existing logic as an API without any behavior change.

sonyps5201314 commented:
> On Windows, the Virtual Memory subsystem won't allocate virtual memory regions at a granularity of less than 64 KBytes of memory anway.

I think that's wrong,In my actual tests, I found that the minimum allocation was 4KB.
```cpp
	PBYTE setup = (PBYTE)detour_alloc_region_from_lo(pbLo, pbHi, 17, TRUE);
	ATLASSERT(setup);
	while (setup< pbLo)
	{
		setup = (PBYTE)detour_alloc_region_from_lo(setup+ 0x10000, pbHi, 17, TRUE);
		MessageBox(NULL,_T("TEST"),_T("APP"),MB_OK);
		if (setup == NULL)
		{
			// Out of memory.
			return NULL;
		}
	}
```
![image](https://user-images.githubusercontent.com/20564099/74106849-3212c780-4ba5-11ea-8703-dff3fd37587a.png)
